workers/generic-worker/os_groups_insecure.go

```diff
@@ -13,11 +15,61 @@ type OSGroups struct {
 }
 
 func (osGroups *OSGroups) Start() *CommandExecutionError {
-	if len(osGroups.Task.Payload.OSGroups) > 0 {
-		return MalformedPayloadError(fmt.Errorf("osGroups feature is not supported on platform %v - please modify task definition and try again", runtime.GOOS))
+	groupNames := osGroups.Task.Payload.OSGroups
+	if len(groupNames) == 0 {
+		return nil
 	}
+
+	userGroups, err := currentUserGroups()
+	if err != nil {
+		return executionError(internalError, errored, fmt.Errorf("failed to get user groups: %v", err))
+	}
+
+	notInGroups := checkUserGroupsInList(groupNames, userGroups)
+	if len(notInGroups) > 0 {
+		return MalformedPayloadError(fmt.Errorf("task payload contains unsupported osGroups: %v\nallowed values (on this worker pool): %v", notInGroups, userGroups))
+	}
+
 	return nil
 }
 
 func (osGroups *OSGroups) Stop(err *ExecutionErrors) {
 }
+
+func currentUserGroups() (map[string]bool, error) {
+	currentUser, err := user.Current()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current user: %v", err)
+	}
+
+	groupIDs, err := currentUser.GroupIds()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get current user's group ids: %v", err)
+	}
+
+	primaryGID := currentUser.Gid
+	userGroups := make(map[string]bool)
+	for _, gid := range groupIDs {
+		if gid == primaryGID {
+			continue
```

Comment:
I don't think this is quite what we want. On some platforms, the group may be fixed (e.g. "staff" on macOS) but on other platforms (e.g. Linux) it may be variable, and match the task user name. I think only in the case it is variable, i.e. matches the user name, we would want to filter it out. So I think here we just need to filter out the case when it matches the username exactly. If it is the primary group, but doesn't match the username, we should keep it.